services/proxy/pkg/middleware/client_api_key_auth.go

```diff
@@ -0,0 +1,136 @@
+package middleware
+
+import (
+	"encoding/json"
+	"errors"
+	revactx "github.com/cs3org/reva/v2/pkg/ctx"
+	"github.com/golang-jwt/jwt"
+	"github.com/google/uuid"
+	"github.com/owncloud/ocis/v2/ocis-pkg/log"
+	"github.com/owncloud/ocis/v2/services/proxy/pkg/user/backend"
+	"go-micro.dev/v4/store"
+	"net/http"
+)
+
+// ClientAPIKeyAuthenticator is the authenticator responsible for client API keys.
+type ClientAPIKeyAuthenticator struct {
+	Logger       log.Logger
+	UserProvider backend.UserBackend
+	SigningKey   string
+	Store        store.Store
+}
+
+type ClientAPIKeyRecord struct {
+	UserId string
+	Scopes []string
+}
+
+func (m ClientAPIKeyAuthenticator) CreateClientAPIKey() (string, string, error) {
+	k := uuid.New()
+	signature, err := jwt.SigningMethodHS256.Sign(k.String(), []byte(m.SigningKey))
+	if err != nil {
+		return "", "", err
+	}
+	return k.String(), signature, nil
+}
+
+func (m ClientAPIKeyAuthenticator) SaveKey(userId string, keyid string) error {
+	record := ClientAPIKeyRecord{
+		UserId: userId,
+	}
+	bytes, err := json.Marshal(record)
+	if err != nil {
+		return err
+	}
+	r := store.Record{
+		Key:   keyid,
+		Value: bytes,
+	}
+	return m.Store.Write(&r)
+}
+
+// Authenticate implements the authenticator interface to authenticate requests via basic auth.
+func (m ClientAPIKeyAuthenticator) Authenticate(r *http.Request) (*http.Request, bool) {
+	if isPublicPath(r.URL.Path) {
+		// The authentication of public path requests is handled by another authenticator.
+		// Since we can't guarantee the order of execution of the authenticators, we better
+		// implement an early return here for paths we can't authenticate in this authenticator.
+		return nil, false
+	}
+
+	clientApiKeyId, clientApiKeySecret, ok := r.BasicAuth()
+	if !ok {
+		return nil, false
+	}
+
+	// re-build JWT
+	err := m.VerifyBasicAuth(clientApiKeyId, clientApiKeySecret)
+	if err != nil {
+		m.Logger.Error().
+			Err(err).
+			Str("authenticator", "client_api_keys").
+			Str("path", r.URL.Path).
+			Msg("failed to authenticate request")
+		return nil, false
+	}
+
+	// lookup user data in micro store
+	clientAPIKeyRecord, err := m.LookupUser(clientApiKeyId)
+	if err != nil {
+		m.Logger.Error().
+			Err(err).
+			Str("authenticator", "client_api_keys").
+			Str("path", r.URL.Path).
+			Msg("failed to lookup client api key record")
+		return nil, false
+	}
+
+	_, token, err := m.UserProvider.GetUserByClaims(r.Context(), "userid", clientAPIKeyRecord.UserId)
```

Comment:
Hm, I'd rather prefer not use the `machine` authprovider to generate the reva token here. Ideally we would have a separate auth manager in reva to handle the api keys.

IMO the  `machine` authprovider is a crutch and we should get away from using it.

BTW there is already the `appauth` manager in reva (https://github.com/cs3org/reva/blob/edge/pkg/auth/manager/appauth/appauth.go) which seems to serve a similar purpose. Maybe we can use that?

Reply:
@butonic @rhafer let's try to find a time slot tomorrow to discuss this ... or when ever time permits ...

Reply:
> BTW there is already the `appauth` manager in reva (https://github.com/cs3org/reva/blob/edge/pkg/auth/manager/appauth/appauth.go) which seems to serve a similar purpose. Maybe we can use that?

there is currently only one implementation which stores the password in a json file - https://github.com/owncloud/ocis/blob/afc6ed1e416c8c1abef832df3e6912b7487cccbd/vendor/github.com/cs3org/reva/v2/pkg/appauth/manager/json/json.go

And ... oh lord ... so many layers :exploding_head: .... what happened to the good old KISS approach ;-)

Reply:
Unless I miss anything: in case the final implementation of this feature lives within reva it requires a grpc call which somehow violates the dos mitigation requirement as described in the ADR #8515 ......